### DISCUSSION TITLE
lspd 中的 finishReceiver 是否不必要？

### DISCUSSION THREAD

**Discussion by 5ec1cff** [LSPosed/LSPosed#2138] [General]
LSPosed 的 daemon 中[注册了一系列 BroadcastReceiver](https://github.com/LSPosed/LSPosed/blob/a0dd9c0866ad744dd089ef44db1825a7e8c39b2b/daemon/src/main/java/org/lsposed/lspd/service/LSPosedService.java#L235) ，并且在调用结束后都会调用 `AMS.finishReceiver` 。

阅读 [Android 源码](https://android.googlesource.com/platform/frameworks/base/+/8ca6061a15e6097e350a5c65385d573e8182c29a/core/java/android/content/BroadcastReceiver.java#283)得知， `finishReceiver` 一般仅当广播接收器收到有序广播 ( `ordered=true` ) 或者广播接收器为 manifest 中声明的时候才调用。事实上，LSPosed 需要接收的广播基本都是无序广播。

尽管对无序广播调用 `finishReceiver` 并不会产生异常，但是在 AMS 中它总会[调用 trimApplication](https://android.googlesource.com/platform/frameworks/base/+/8ca6061a15e6097e350a5c65385d573e8182c29a/services/core/java/com/android/server/am/ActivityManagerService.java#14524)。在 Android 11 及以下，这会[更新所有进程的 OomAdj](https://android.googlesource.com/platform/frameworks/base/+/34a1b9c951c38537ab96b69bc308f6e0884823f5/services/core/java/com/android/server/am/ActivityManagerService.java#18409) ，这个过程可能导致卡顿。从个人体验来看，安装了 LSPosed 的情况下，开机的时候卡顿尤其明显，推测是由于此阶段 LSPosed 接收了大量广播，导致 finishReceiver 被频繁调用。

尝试修改 LSPosed 源码，阻止对无序广播调用 finishReceiver ，实测发现不影响正常功能，而开机卡顿感明显减弱。

因此想询问开发者，这个调用是否是不必要的？

**Comment by yujincheng08** [comment]
#684

**Reply by 5ec1cff** [reply]
> #684

看上去 [`ACTION_LOCKED_BOOT_COMPLETED`](https://android.googlesource.com/platform/frameworks/base/+/13db8fd5707b3c1c1484d717da8bda73ee7090a6/services/core/java/com/android/server/am/UserController.java#621) 是有序广播，不过 [`ACTION_CONFIGURATION_CHANGED`](https://android.googlesource.com/platform/frameworks/base/+/13db8fd5707b3c1c1484d717da8bda73ee7090a6/services/core/java/com/android/server/am/ActivityManagerService.java#17104) [`ACTION_PACKAGE_*`](https://android.googlesource.com/platform/frameworks/base/+/13db8fd5707b3c1c1484d717da8bda73ee7090a6/services/core/java/com/android/server/pm/BroadcastHelper.java) 一般是无序广播。

因此我认为改成下面这样没有问题：

```java
if (ordered) {
    ActivityManagerService.finishReceiver(this, resultCode, data, extras, false, intent.getFlags());
}
```

**Reply by yujincheng08** [reply]
欢迎 pr

**Reply by 5ec1cff** [reply]
谢谢西大师！
